File: 2015/Rust/aoc_2015/src/day_3.rs
use std::fs;

pub fn day_3_main(){
    part_1();
    part_2();
}

fn part_1(){
    //Ingest data
    let file_name = "inputs/day_3_input.txt";
    let contents = fs::read_to_string(file_name)
        .expect("Failed to read File");

    let parts: Vec<&str> = contents.split("").collect();

    let mut locmem: Vec<[i32; 2]> = Vec::new();
    let mut loc: [i32; 2] = [0, 0];

    for i in 0..parts.len()-2{
        if parts[i] == "^"{
            loc[1] += 1;
        }else if parts[i] == "v"{
            loc[1] -= 1;
        }else if parts[i] == "<"{
            loc[0] -= 1;
        }else if parts[i] == ">"{
            loc[0] += 1;
        }

        if locmem.contains(&loc) == false{
            locmem.push(loc);
        }
    }

    println!("Total Houses seen in Part 1: {}", locmem.len());
}

fn part_2(){
    //Ingest data
    let file_name = "inputs/day_3_input.txt";
    let contents = fs::read_to_string(file_name)
        .expect("Failed to read File");

    let parts: Vec<&str> = contents.split("").collect();

    let mut locmem: Vec<[i32; 2]> = Vec::new();
    let mut loc_santa: [i32; 2] = [0, 0];
    let mut loc_robo_santa: [i32; 2] = [0, 0];
    let mut turn: i32 = 0;

    for i in 0..parts.len()-2{
        if parts[i] == "^"{
            if turn == 0{
                loc_santa[1] += 1
            }else{
                loc_robo_santa[1] += 1;
            }
        }else if parts[i] == "v"{
            if turn == 0{
                loc_santa[1] -= 1;
            }else{
                loc_robo_santa[1] -= 1;
            }
        }else if parts[i] == "<"{
            if turn == 0{
                loc_santa[0] -= 1;
            }else{
                loc_robo_santa[0] -= 1;
            }
        }else if parts[i] == ">"{
            if turn == 0{
                loc_santa[0] += 1;
            }else{
                loc_robo_santa[0] += 1;
            }
        }


        if turn == 0{
            if locmem.contains(&loc_santa) == false{
                locmem.push(loc_santa);
            }
            turn = 1;
        }else{
            if locmem.contains(&loc_robo_santa) == false{
                locmem.push(loc_robo_santa);
            }
            turn = 0;
        }
    }

    println!("Total Houses seen in Part 2: {}", locmem.len());
}

File: 2015/Python/Day_3.py
def parse_input():
    fullArray = []

    with open("Inputs/day_3_input.txt") as input_file:
        for line in input_file:
            fullArray = list(line)
    return fullArray


def part_1(list_1):
    loc = [0, 0]
    loc_mem = []

    for i in list_1:
        if i == "^":
            loc[1] += 1
        elif i == "v":
            loc[1] -= 1
        elif i == "<":
            loc[0] -= 1
        elif i == ">":
            loc[0] += 1

        if loc in loc_mem:
            pass
        else:
            loc_mem.append(loc.copy())

    return len(loc_mem) + 1


def part_2(list_1):
    loc_mem = []
    loc_santa = [0, 0]
    loc_robo = [0, 0]
    turn = 0

    for i in range(len(list_1)):
        if list_1[i] == "^":
            if turn == 0:
                loc_santa[1] += 1
            else:
                loc_robo[1] += 1
        elif list_1[i] == "v":
            if turn == 0:
                loc_santa[1] -= 1
            else:
                loc_robo[1] -= 1
        elif list_1[i] == "<":
            if turn == 0:
                loc_santa[0] -= 1
            else:
                loc_robo[0] -= 1
        elif list_1[i] == ">":
            if turn == 0:
                loc_santa[0] += 1
            else:
                loc_robo[0] += 1

        if turn == 0:
            if loc_santa in loc_mem:
                pass
            else:
                loc_mem.append(loc_santa.copy())
            turn = 1
        else:
            if loc_robo in loc_mem:
                pass
            else:
                loc_mem.append(loc_robo.copy())
            turn = 0

    return len(loc_mem)


if __name__ == "__main__":
    list_1 = []

    list_1 = parse_input()

    print("The answer for Part 1: " + str(part_1(list_1)))
    print("The answer for Part 2: " + str(part_2(list_1)))
